File: packages/simple-cli/src/repl.ts
import readline from "readline/promises"
import { stdin as input, stdout as output } from "process"
import { Renderer } from "./renderer"
import { applyCommand, helpText, parseCommand } from "./commands"
import type { ResolvedConfig } from "./types"
import type { SimpleClient } from "./client"

function buildPrompt(): string {
  const cwd = process.cwd()
  const dirName = cwd.split(/\\|\//).pop() || cwd
  return `${dirName}> `
}

async function readMultiline(rl: readline.Interface): Promise<string | null> {
  let buffer = ""
  while (true) {
    const prompt = buffer ? "... " : buildPrompt()
    const line = await rl.question(prompt).catch(() => null)
    if (line === null) return null
    if (line.endsWith("\\")) {
      buffer += line.slice(0, -1)
      buffer += "\n"
      continue
    }
    buffer += line
    return buffer
  }
}

export async function runRepl(config: ResolvedConfig, client: SimpleClient, renderer: Renderer) {
  const rl = readline.createInterface({ input, output })
  renderer.banner(config.url)

  while (true) {
    const line = await readMultiline(rl)
    if (line === null) break
    const trimmed = line.trim()
    if (!trimmed) continue

    if (trimmed.startsWith("/")) {
      const cmd = parseCommand(trimmed)
      switch (cmd.type) {
        case "exit":
          rl.close()
          client.close()
          return
        case "help":
          renderer.help(helpText())
          continue
        case "set":
          const next = applyCommand(cmd, config)
          Object.assign(config, next)
          renderer.trace("updated", { [cmd.key]: cmd.value })
          continue
        case "unknown":
          renderer.help(`Unknown command: ${cmd.input}\n${helpText()}`)
          continue
      }
    }

    renderer.user(trimmed)
    try {
      const response = await client.sendPrompt(trimmed)
      renderer.renderMessage(response.info, response.parts)
    } catch (err) {
      renderer.trace("prompt failed", { error: String(err) })
      console.error("Failed to send prompt:", err)
    }
  }
}

File: go-cli/repl.go
package main

import (
    "bufio"
    "context"
    "fmt"
    "os"
    "strings"
)

func buildPrompt() string {
    cwd, _ := os.Getwd()
    parts := strings.Split(cwd, string(os.PathSeparator))
    last := parts[len(parts)-1]
    if last == "" {
        last = cwd
    }
    return fmt.Sprintf("%s> ", last)
}

func readMultiline(reader *bufio.Reader) (string, error) {
    var lines []string
    for {
        prompt := buildPrompt()
        if len(lines) > 0 {
            prompt = "... "
        }
        fmt.Print(prompt)
        line, err := reader.ReadString('\n')
        if err != nil {
            if len(lines) == 0 {
                return "", err
            }
            return strings.Join(lines, "\n"), nil
        }
        line = strings.TrimRight(line, "\r\n")
        if strings.HasSuffix(line, "\\") {
            lines = append(lines, strings.TrimSuffix(line, "\\"))
            continue
        }
        lines = append(lines, line)
        return strings.Join(lines, "\n"), nil
    }
}

func runRepl(cfg *ResolvedConfig, client *SimpleClient, renderer *Renderer) error {
    renderer.Banner(cfg.URL)
    reader := bufio.NewReader(os.Stdin)

    for {
        line, err := readMultiline(reader)
        if err != nil {
            return err
        }
        trimmed := strings.TrimSpace(line)
        if trimmed == "" {
            continue
        }

        if strings.HasPrefix(trimmed, "/") {
            cmd := parseCommand(trimmed)
            switch cmd.Type {
            case "exit":
                client.Close()
                return nil
            case "help":
                renderer.Help(helpText)
                continue
            case "set":
                applyCommand(cfg, cmd)
                renderer.Trace("updated", map[string]any{cmd.Key: cmd.Val})
                continue
            default:
                renderer.Help(fmt.Sprintf("Unknown command: %s\n%s", cmd.Val, helpText))
                continue
            }
        }

        renderer.User(trimmed)
        resp, err := client.SendPrompt(context.Background(), trimmed, *cfg)
        if err != nil {
            renderer.Trace("prompt failed", map[string]any{"error": err.Error()})
            fmt.Fprintf(os.Stderr, "Failed to send prompt: %v\n", err)
            continue
        }
        renderer.RenderMessage(resp.Info.ID, true, resp.Parts)
    }
}
